pandas/_testing/asserters.py

```diff
@@ -367,8 +367,8 @@ def _get_ilevel_values(index, level):
 
     # If order doesn't matter then sort the index entries
     if not check_order:
-        left = Index(safe_sort(left))
-        right = Index(safe_sort(right))
+        left = Index(safe_sort(left), dtype=left.dtype)
+        right = Index(safe_sort(right), dtype=right.dtype)
```

Comment:
maybe another issue, but the check_names check is later. It maybe that would need to resuse the code pattern you added in #47206. Not here, as this PR is targeted 1.4.3, just an observation.

Reply:
Good point, I am not using that myself, so I am a bit fuzzy on all the keywords :)

It is actually a bit more comlex, because These calls return regular indexes, even if a MultiIndex was given. I'll open an issue about it

Reply:
my thinking was that your fix in #47206 looked quite useful to robustly sort indexes. and maybe a safe_sort_index would be handy. (IIRC we cannot support pandas objects in safe_sort but would need to check this). I wonder can the fix in https://github.com/pandas-dev/pandas/pull/47206 suffer from the rogue warning too without the dtype specified there?

Reply:
Yep you are completely right. Opened #47330 for 1.5